hack/oci-pin/internal/sums/sums.go

```diff
@@ -0,0 +1,225 @@
+		if refs[i].Registry != refs[j].Registry {
+			return refs[i].Registry < refs[j].Registry
+		}
+		if refs[i].Prefix != refs[j].Prefix {
+			return refs[i].Prefix < refs[j].Prefix
+		}
+		if refs[i].Name != refs[j].Name {
+			return refs[i].Name < refs[j].Name
+		}
+		if refs[i].Tag != refs[j].Tag {
+			return refs[i].Tag < refs[j].Tag
+		}
+		return refs[i].Digest < refs[j].Digest
+	})
+}
+
+func compareRefs(a, b PinnedImageReference) (equal bool, err error) {
+	if a.Registry != b.Registry {
+		return false, nil
+	}
+	if a.Prefix != b.Prefix {
+		return false, nil
+	}
+	if a.Name != b.Name {
+		return false, nil
+	}
+	if a.Tag != b.Tag {
+		return false, nil
+	}
+	if a.Digest != b.Digest {
+		return false, errors.New("conflicting digests")
+	}
+	return true, nil
+}
+
+func parseLines(scanner scanner) ([]PinnedImageReference, error) {
+	var refs []PinnedImageReference
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == "" {
+			continue
+		}
+		ref, err := parseLine(line)
+		if err != nil {
+			return nil, err
+		}
+		refs = append(refs, ref)
+	}
+	return refs, nil
+}
+
+// parseLine parses a line from a sha256sums file.
+// The line is expected to be in the format "hex  reference".
+func parseLine(line string) (PinnedImageReference, error) {
+	if len(line) < minLineLength {
+		return PinnedImageReference{}, fmt.Errorf("line too short: %q", line)
+	}
+	sum := line[:64]
+	ref := line[66:]
+	return refFromSumAndRef(sum, ref)
+}
+
+func refFromSumAndRef(sum, ref string) (PinnedImageReference, error) {
+	if !sumRegexp.MatchString(sum) {
+		return PinnedImageReference{}, fmt.Errorf("invalid sum: %q", sum)
+	}
+
+	// last colon is separator between name and tag
+	var base, tag string
+	tagSep := strings.LastIndexByte(ref, ':')
+	if tagSep == notFound {
+		base = ref
+	} else {
+		base = ref[:tagSep]
+		tag = ref[tagSep+1:]
+	}
+
+	// first slash is separator between registry and full name
+	registrySep := strings.IndexByte(base, '/')
+	if registrySep == notFound {
+		return PinnedImageReference{}, fmt.Errorf("invalid reference: missing registry %q", ref)
+	}
+
+	registry := base[:registrySep]
+	fullName := base[registrySep+1:]
+
+	// last slash is separator between prefix and short name
+	var prefix, name string
+	nameSep := strings.LastIndexByte(fullName, '/')
+	if nameSep == notFound {
+		name = fullName
+	} else {
+		prefix = fullName[:nameSep]
+		name = fullName[nameSep+1:]
+	}
+
+	return PinnedImageReference{
+		Registry: registry,
+		Prefix:   prefix,
+		Name:     name,
+		Tag:      tag,
+		Digest:   "sha256:" + sum,
+	}, nil
+}
+
+func newLineScanner(r io.Reader) scanner {
+	scanner := bufio.NewScanner(r)
+	scanner.Split(bufio.ScanLines)
+	return scanner
+}
+
+type scanner interface {
+	Scan() bool
+	Text() string
+}
+
+var sumRegexp = regexp.MustCompile(`^[a-f0-9]{64}$`)
+
+const (
+	sumLength     = 64
+	minLineLength = sumLength + 2 // 64 hex + 2 space
+	notFound      = -1            // index not found
+)
```

Comment:
I personally like to have the constants at the top of the file. This aligns it with e.g. the Regexes in the cloud metadata.